src/template_finder.py

```diff
@@ -63,9 +70,98 @@ def __new__(cls, save_last_res=False):
+            normalize_monitor: bool = False,
+            use_grayscale: bool = False,
+            color_match: list = False,
+    ) -> list[TemplateMatch]:
+        """Look for matches for a template inside an image.
+
+        :param image: The image to find the template in.
+        :param template_name: The name of the template to load to look for in the image.
+        :param template_image: An already loaded template to look for in the image.
+        :param threshold: The confidence threshold to consider a match valid.
+        :param roi: Optional region of the image to look for templates in.
+        :param normalize_monitor: Return monitor coordinates instead of input image coordinates.
+        :param use_grayscale: Whether the match should be in grayscale.
+        :param color_match: Pass a color to be used as a filter for the template and image.
+
+        :returns results: All templates found in the specified region.
+        """
+        results = []
+        if template_name:
+            scale = self._templates[template_name][2]
+            mask = self._templates[template_name][3]
+            if color_match:
+                templates = [color_filter(self._templates[template_name][0], color_match)[1]]
+                image = color_filter(image, color_match)[1]
+            else:
+                templates = [self._templates[template_name][use_grayscale]]
+        elif template_image is not None:
+            if not color_match:
+                templates = [template_image]
+            else:
+                templates = [color_filter(template_image, color_match)[1]]
+            scale = 1.0
+            mask = None
+        else:
+            raise ValueError('We need either a template_name or template_image to search for.')
+        if use_grayscale:
+            image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+            templates = [cv2.cvtColor(i, cv2.COLOR_BGR2GRAY) for i in templates]
+        # Default ROI to full image if not specified.
+        roi = roi if roi is not None else [0, 0, image.shape[1], image.shape[0]]
+        if scale != 1:
+            image: np.ndarray = cv2.resize(image, None, fx=scale, fy=scale, interpolation=cv2.INTER_NEAREST)
+            roi = [val * scale for val in roi]
+        # Crop image to region of interest.
+        rx, ry, rw, rh = roi
+        image = image[ry:ry + rh, rx:rx + rw]
+
+        for idx, template in enumerate(templates):
+            if image.shape[0] > template.shape[0] and image.shape[1] > template.shape[1]:
+                res = cv2.matchTemplate(image, template, cv2.TM_CCOEFF_NORMED, mask=mask)
+                res = np.nan_to_num(res, copy=False, nan=0.0, posinf=0.0, neginf=0.0)
+                matches = np.where(res >= threshold)
+                matches = [pos for pos in zip(*matches)]
+                for coords in matches:
+                    # TODO: The TemplateMatch class should be able to construct itself with this information.
```

Comment:
I'm still not super familiar with the codebase but I believe this assessment to be true.
I know it's a data class but if there's always logic associates with it maybe it should just be a normal object. Or not, who knows, lol.